ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables/UtxoDiff.hs

```diff
@@ -0,0 +1,177 @@
+
+{------------------------------------------------------------------------------
+  Construction
+------------------------------------------------------------------------------}
+
+diff :: Map k v -> Map k v -> UtxoDiff k v
+diff = UtxoDiff
+
+fromMapInserts :: Ord k => Map k v -> UtxoDiff k v
+fromMapInserts = UtxoDiff mempty
+
+fromMapDeletes :: Ord k => Map k v -> UtxoDiff k v
+fromMapDeletes = flip UtxoDiff mempty
+
+fromListInserts :: Ord k => [(k, v)] -> UtxoDiff k v
+fromListInserts = UtxoDiff mempty . Map.fromList
+
+fromListDeletes :: Ord k => [(k, v)] -> UtxoDiff k v
+fromListDeletes = flip UtxoDiff mempty . Map.fromList
+
+{------------------------------------------------------------------------------
+  Query
+------------------------------------------------------------------------------}
+
+null :: UtxoDiff k v -> Bool
+null (UtxoDiff m1 m2) = Map.null m1 && Map.null m2
+
+size :: UtxoDiff k v -> Int
+size (UtxoDiff m1 m2) = Map.size m1 + Map.size m2
+
+numInserts :: UtxoDiff k v -> Int
+numInserts (UtxoDiff _ m2) = Map.size m2
+
+numDeletes :: UtxoDiff k v -> Int
+numDeletes (UtxoDiff m1 _) = Map.size m1
+
+{------------------------------------------------------------------------------
+  Applying diffs
+------------------------------------------------------------------------------}
+
+applyUtxoDiff ::
+     Ord k
+  => Map k v
+  -> UtxoDiff k v
+  -> Map k v
+applyUtxoDiff m (UtxoDiff deletes inserts) =
+    (m `Map.difference` deletes) `Map.union` inserts
+
+applyUtxoDiffForKeys ::
+     Ord k
+  => Map k v
+  -> Set k
+  -> UtxoDiff k v
+  -> Map k v
+applyUtxoDiffForKeys m ks (UtxoDiff deletes inserts) =
+  applyUtxoDiff
+    m
+    (UtxoDiff (deletes `Map.restrictKeys` (Map.keysSet m `Set.union` ks))
+          (inserts `Map.restrictKeys` (Map.keysSet m `Set.union` ks)))
+
+{-------------------------------------------------------------------------------
+  Filter
+-------------------------------------------------------------------------------}
+
+filterOnlyKey :: (k -> Bool) -> UtxoDiff k v -> UtxoDiff k v
+filterOnlyKey f (UtxoDiff m1 m2) =
+  UtxoDiff (Map.filterWithKey (const . f) m1) (Map.filterWithKey (const . f) m2)
+
+{-------------------------------------------------------------------------------
+  Reductive instances
+-------------------------------------------------------------------------------}
+
+instance (Ord k, Eq v) => LeftReductive (UtxoDiff k v) where
+  stripPrefix (UtxoDiff m1 m2) (UtxoDiff m3 m4) =
+    let
+        -- m1 ⊆ m3 because of UTXO-property
+        -- we cannot have added something that was deleted
+        dels = stripPrefix m1 m3
+    in
+      case (dels, stripPrefix m2 m4) of
+        -- Violation of UTXO-property
+        (Nothing, _) -> Nothing
+        -- m2 ⊄ m4 because some have been cancelled
+        (Just dels', Nothing) ->
+          let
+            (notCancelled, cancelled) = Map.partitionWithKey (\k _ -> k `Map.member` m4) m2
```

Comment:
This looks like it could be a rather costly operation: doing a `Map.member` for each key. Maybe using `Map.difference` or something like that is cheaper